### ISSUE TITLE
[Windows-only] Vieb does not fully quit on Windows because deleting files freezes the entire app

### ISSUE THREAD

**Issue by Melandel** [Jelmerro/Vieb#526]
**Checklist**
- [x] I searched the `:help` documentation using `/` for helpful information
- [x] I have read the [frequently asked questions](https://github.com/Jelmerro/Vieb/blob/master/FAQ.md)
- [x] I did not find any similar issues in the [issue tracker](https://github.com/Jelmerro/Vieb/issues)
- [x] I prefer making an issue instead of using [discussions](https://github.com/Jelmerro/Vieb/discussions), [Matrix](https://matrix.to/#/#vieb:matrix.org) or [Telegram](https://t.me/vieb_general)

**Describe the bug**
Running Vieb, quitting it, then re-running it from the command line:
* ends up: displaying `Sending urls to existing instance C:\Users\Foo\Bar\Baz\viebfiles/` in the command line
* instead of: opening Vieb's graphical interface

**To Reproduce**
* Open command line and type `Vieb.exe "https://github.com/"`
* In Vieb, type `:quitall`
* In command line, type `Vieb.exe "https://www.wikipedia.org/"`

**Expected behavior**
Vieb is open on the wikipedia page

**Specs**
 - OS: Windows 10
 - Version: Vieb 10.4.0

**Additional information**
Apparently, there are Vieb processes still running after quitting vieb:
![image](https://github.com/Jelmerro/Vieb/assets/48007894/55b9a316-ab9a-4cce-8a1c-0b7417c7ec38)

**Comment by Jelmerro** [comment]
Not quitting correctly with `:quitall` can be debugged by starting Vieb with `--devtools` the first time and then doing the `:quitall`, the window should then stay open and show the error in the devtools. Please share it here so I can know the issue. Is this new in 10.4.0 compared to older versions?

**Comment by Melandel** [comment]
It looks like deleting the `viebfiles` folder and rebooting Windows got rid of the issue - I cannot reproduce the bug.

**Comment by Jelmerro** [comment]
Can you reproduce this with a clean config? Because I cannot reproduce this on Windows 11 at all. And how about a separate datafolder? If there's no error on quit and it keeps running, even when running with `--devtools`, it either freezes at Chromium level or keeps running one of the `while` loops in the code indefinitely, of which there are not a lot:
https://github.com/search?q=repo%3AJelmerro%2FVieb%20while&type=code
And the only one that could possibly run on quit is the rimraf one, which stops after some retries. Let me know if you have any other info, because I cannot reproduce or know the issue if it freezes without errors.

**Comment by Melandel** [comment]
> Can you reproduce this with a clean config? How about a separate datafolder?

I am able to confirm that by removing my `VIEB_CONFIG_FILE` and `VIEB_DATAFOLDER` environment parameters, I am able to reproduce the bug on Windows 10 using `npm ci; npm start "https://www.bienici.com/recherche/location/paris-75000/2-pieces-et-plus?prix-max=900&chambres-min=1"`

I tried adding some `console.log` inside the function `quitall`, but they did not show on my terminal.

I tried commenting instructions, and was able to investigate further using that method.

I am able to confirm that commenting out these 2 lines exclusively([here](https://github.com/Jelmerro/Vieb/blob/67ef838ffd45727cfa84aace782af512355d1865/app/util.js#L1492) and [here](https://github.com/Jelmerro/Vieb/blob/67ef838ffd45727cfa84aace782af512355d1865/app/util.js#L1495)) removes the freezing, somehow

**Comment by Jelmerro** [comment]
So it's basically either Windows that blocks the deletion of some files, or an infinite loop in this script: https://github.com/Jelmerro/Vieb/blob/master/app/rimraf.js Could you figure out where exactly it freezes? Because even after building up some history and files in those directories I managed to quit the browser without error on Windows 10 too.

**Comment by Melandel** [comment]
Thanks for the `F8` tip!

### Logs

<details>
  <summary>Click me!</summary>

```text
Vieb\app\rimraf.js:136 [C:\Users\ANON\AppData\Roaming\Vieb\Service Worker] Trying to fetch statistics
Vieb\app\rimraf.js:144 [C:\Users\ANON\AppData\Roaming\Vieb\Service Worker] 🚫 Statistics could not be fetched {
  "errno": -4058,
  "syscall": "lstat",
  "code": "ENOENT",
  "path": "C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\Service Worker"
}
Vieb\app\rimraf.js:148 [C:\Users\ANON\AppData\Roaming\Vieb\Service Worker] ✅ Abort (ENOENT means "file not found")
Vieb\app\rimraf.js:136 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] Trying to fetch statistics
Vieb\app\rimraf.js:140 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] ✅ Statistics fetched successfully
Vieb\app\rimraf.js:164 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] Trying to remove directory
Vieb\app\rimraf.js:44 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] Trying to call fs.rmdirSync
Vieb\app\rimraf.js:48 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] ✅ fs.rmdirSync was successful
Vieb\app\rimraf.js:168 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\Service Worker] ✅ Directory removed successfully
Vieb\app\rimraf.js:77 [C:\Users\ANON\AppData\Roaming\Vieb\blob_storage] Now starting internal fs.rmSync retry strategy (100 retries)
Vieb\app\rimraf.js:80 [C:\Users\ANON\AppData\Roaming\Vieb\blob_storage] internal fs.rmSync retry #1 on 100...
Vieb\app\rimraf.js:84 [C:\Users\ANON\AppData\Roaming\Vieb\blob_storage] ✅ internal fs.rmSync retry #1 was successful
Vieb\app\rimraf.js:77 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\blob_storage] Now starting internal fs.rmSync retry strategy (100 retries)
Vieb\app\rimraf.js:80 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\blob_storage] internal fs.rmSync retry #1 on 100...
Vieb\app\rimraf.js:84 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\blob_storage] ✅ internal fs.rmSync retry #1 was successful
Vieb\app\rimraf.js:136 [C:\Users\ANON\AppData\Roaming\Vieb\databases] Trying to fetch statistics
Vieb\app\rimraf.js:144 [C:\Users\ANON\AppData\Roaming\Vieb\databases] 🚫 Statistics could not be fetched {
  "errno": -4058,
  "syscall": "lstat",
  "code": "ENOENT",
  "path": "C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\databases"
}
Vieb\app\rimraf.js:148 [C:\Users\ANON\AppData\Roaming\Vieb\databases] ✅ Abort (ENOENT means "file not found")
Vieb\app\rimraf.js:136 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] Trying to fetch statistics
Vieb\app\rimraf.js:140 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] ✅ Statistics fetched successfully
Vieb\app\rimraf.js:164 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] Trying to remove directory
Vieb\app\rimraf.js:44 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] Trying to call fs.rmdirSync
Vieb\app\rimraf.js:52 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] 🚫 fs.rmdirSync failed {
  "errno": -4051,
  "syscall": "rmdir",
  "code": "ENOTEMPTY",
  "path": "C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\Partitions\\main\\databases"
}
Vieb\app\rimraf.js:136 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] Trying to fetch statistics
Vieb\app\rimraf.js:140 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] ✅ Statistics fetched successfully
Vieb\app\rimraf.js:172 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] Trying to unlink file
Vieb\app\rimraf.js:184 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] 🚫 File unlink failed {
  "errno": -4082,
  "syscall": "unlink",
  "code": "EBUSY",
  "path": "C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\Partitions\\main\\databases\\Databases.db"
}
Vieb\app\rimraf.js:212 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] Calling rmSync
Vieb\app\rimraf.js:44 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] Trying to call fs.rmdirSync
Vieb\app\rimraf.js:52 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] 🚫 fs.rmdirSync failed {
  "errno": -4058,
  "syscall": "rmdir",
  "code": "ENOENT",
  "path": "C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\Partitions\\main\\databases\\Databases.db"
}
Vieb\app\rimraf.js:56 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases\Databases.db] Abort (ENOENT means "file not found")
Vieb\app\rimraf.js:77 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] Now starting internal fs.rmSync retry strategy (100 retries)
Vieb\app\rimraf.js:80 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] internal fs.rmSync retry #1 on 100...
Vieb\app\rimraf.js:87 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] 🚫 internal fs.rmSync retry #1 failed {
  "errno": -4082,
  "syscall": "unlink",
  "code": "EBUSY",
  "path": "\\\\?\\C:\\Users\\ANON\\AppData\\Roaming\\Vieb\\Partitions\\main\\databases\\Databases.db"
}
Vieb\app\rimraf.js:80 [C:\Users\ANON\AppData\Roaming\Vieb\Partitions\main\databases] internal fs.rmSync retry #2 on 100...
```

</details>

### Analysis

It looks like there is a mix of 3 things:
1. calling `fs.rmSync(p, { "maxRetries": 100, "recursive": true })` on a busy (`EBUSY` error) resource loops infinitely
2. After failing to unlink a file found busy (`EBUSY` error), calling `fs.rmdirSync` wrongfully find it inexistant (`ENOENT` error) and stops the removal attempt
3. Why is `rimraf` trying to delete a `directory` (`fs.rmdirSync`) on a file after failing to unlink it?

### Additional information

I've been playing the scenario multiple times, and:
* the amount of retries vary from run to run
* the files that are found busy vary from run to run
* but the 3 points I stated above always stand true

**Comment by Jelmerro** [comment]
This proves to me what the author of Rimraf (see screenshot) and my docs also explain in the rimraf file: Windows file access is fundamentally broken. The reason it is retried a 100 times only on Windows is because on mac or linux if it fails there's a good reason for it, so 1 retry is more than enough, while Windows keeps the file open for too long and thus breaks file removal. If it freezes on any of the `fs.*` functions, that means that the NodeJS implementation of the filesystem access freezes. This is not something I can solve in Vieb I'm afraid. I am unsure why this is new behavior, as the commit you reference did not change the rimraf file at all, nor change any of the invocations. So either that is not the right commit, or this was not caused by a change in Vieb directly (though possibly it is related to the new NodeJS Chromium and Electron versions that V10 uses compared to 9.7.1). As such, I do not know what I can do to debug or solve this, as working file removal sounds like a fairly basic requirement for having a working OS. I can try to switch back to the more bloated original rimraf module to hope that they have found a fix for this, but I'm skeptical that this is fixable.

![image](https://github.com/Jelmerro/Vieb/assets/1696674/debb5d8e-e065-4300-b82f-a698c0aeeb71)

**Comment by Melandel** [comment]
I understand.

May something like [this](https://stackoverflow.com/a/37707620) help?

**Comment by Jelmerro** [comment]
Very glad that you found a cause for this in the rimraf file, something must have happened to the Windows handling recently, that still works for mac/linux, as I don't have the ability to test Windows regularly. I'll send you a bunch of different versions of the rimraf file soon to test so we might be able to know what caused this. I'll let you know when I have them ready.

**Comment by Jelmerro** [comment]
Actually, before checking anything else, I found a potential problem while comparing the diffs. Can you try replacing this at the end of the file of the V10 version:
```diff
--- a/app/rimraf.js
+++ b/app/rimraf.js
@@ -138,7 +138,7 @@ const rimrafSync = p => {
             }
             return rmdirSync(p, er)
         }
-        if (isErrnoException(er) && er.code === "EISDIR") {
+        if (!isErrnoException(er) || er.code !== "EISDIR") {
             throw er
         }
         if (er instanceof Error) {

````
Please let me know if this changes the behavior, if not, I'll still make the different versions of the rimraf file (or just a check which of these commits is the culprit would be great too: https://github.com/Jelmerro/Vieb/commits/master/app/rimraf.js). If the above fixes it, I imagine this would be the problematic commit: https://github.com/Jelmerro/Vieb/commit/ce658e7a506850e50f120b58782a30386b3807d0

**Comment by Melandel** [comment]
The modification you suggested appears to prevent the freeze from happening  👍

**Comment by Melandel** [comment]
The modification you suggested appears to prevent the freeze from happening 👍

**Comment by Jelmerro** [comment]
I'll include that change in the next release, thanks for all the help!

**Comment by Melandel** [comment]
Thanks for your time and energy always 🥇

**Comment by Jelmerro** [comment]
Fixed in 10.5.0 :tada:
